In [1]:
import pandas as pd

In [2]:
path1 = "../test_library_info/math_department1_test_knowledge_trees.csv"
df_knowledge_trees = pd.read_csv(path1)

In [3]:
path2 = "../test_library_info/test_info.csv"
df_test_info = pd.read_csv(path2)

/Users/mayanli/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (21,32,38,40,44,50,54) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [4]:
df_knowledge_trees.head()

id   name alias_name  description  department  subject_id subject_name  \
0     2  小数知识树        NaN          NaN           1           2           数学   
1  1173   课外数学        NaN          NaN           1           2           数学   
2  1174     计算        NaN          NaN           1           2           数学   
3  1176    应用题        NaN          NaN           1           2           数学   
4  1183     小数        NaN          NaN           1           2           数学   

   parent_id  rank         rank_string  
0          0     1                 .2.  
1          2     2            .2.1173.  
2       1173     3       .2.1173.1174.  
3       1173     3       .2.1173.1176.  
4       1174     4  .2.1173.1174.1183.

In [5]:
df_test_info.head()

ability_level_id  accuracy  accuracy_sort  adding_mode  answer_times  \
0                 0         0              0            2             0   
1                 0         0              0            2             0   
2                 0         0              0            2             0   
3                 0         0              0            2             0   
4                 0         0              0            2             0   

      batch_num  category_id  category_tree_id  choice_type  choose_sort  ...  \
0  2.018071e+13          136               NaN            0            1  ...   
1  2.018071e+13          136               NaN            0            1  ...   
2  2.018071e+13          136               NaN            0            1  ...   
3  2.018071e+13          138               NaN            0            1  ...   
4  2.018071e+13          138               NaN            0            1  ...   

   test_type                                      url use_env used_exam_names  \
0          3  1093993,1093994,1094801,1094804,1141904       1             NaN   
1          3                                      NaN       1             NaN   
2          3                                      NaN       1             NaN   
3          1                                      NaN       1             NaN   
4          1                                      NaN       1             NaN   

   used_handout_names used_sort  used_times          verify_time  \
0                 NaN        10           0  2018-08-03 17:23:39   
1                 NaN        10           0  2018-08-03 17:23:39   
2                 NaN        10           0  2018-08-03 17:23:39   
3                 NaN        10           0  2018-08-03 17:23:39   
4                 NaN        10           0  2018-08-03 17:23:39   

   verify_timestamp  year_sort  
0        1533288219          9  
1        1533288219          9  
2        1533288219          9  
3        1533288219          9  
4        1533288219          9  

[5 rows x 63 columns]

#  test_info题库knowledge_ids分析

In [6]:
df_test_info[["knowledge_ids"]].count()

knowledge_ids    202509
dtype: int64

In [11]:
df_test_info[["knowledge_ids"]].describe()

knowledge_ids
count                 202509
unique                  3564
top     ,27447,27509,2,27672
freq                     999

test_info 一共             (231891, 63)   
knowledge_ids字段不为空     (202509, 63)    


27447	小学综合口算（手写专用 勿标 不能下载）    (194817, 63)    
13796	人教新课标课内知识树（旧版本章节树）      (3783, 63)    
32754	课内数学体系                          (2917, 63)    
1173	课外数学                             (2764, 63)   
32753	计算树3.0人教版（勿标  可下载）         (388, 63)   
13241	初中衔接            (41, 63)    
14191	小高计算            (13, 63)    
13293	小低口算题          (4, 63)   
34731	小学启航年级讲次树    (1, 63)    
35878	北师版知识树         (0, 63)   

In [46]:
df_test_info.shape

(231891, 63)

In [32]:
df_test_info1 = df_test_info[- df_test_info["knowledge_ids"].isnull()]

In [33]:
df_test_info1.shape

(202509, 63)

In [35]:
df_test_info2 = df_test_info1[df_test_info1["knowledge_ids"].str.contains("2")]
df_test_info2.shape

(202509, 63)

In [36]:
df_test_info2 = df_test_info1[df_test_info1["knowledge_ids"].str.contains("1173")]
df_test_info2.shape
# 1173	课外数学

(2764, 63)

In [37]:
df_test_info2 = df_test_info1[df_test_info1["knowledge_ids"].str.contains("13241")]
df_test_info2.shape

(41, 63)

In [39]:
df_test_info2 = df_test_info1[df_test_info1["knowledge_ids"].str.contains("13796")]
df_test_info2.shape
# 13796	人教新课标课内知识树（旧版本章节树）

(3783, 63)

In [40]:
df_test_info2 = df_test_info1[df_test_info1["knowledge_ids"].str.contains("14191")]
df_test_info2.shape

(13, 63)

In [41]:
df_test_info2 = df_test_info1[df_test_info1["knowledge_ids"].str.contains("27447")]
df_test_info2.shape

# 27447	小学综合口算（手写专用 勿标 不能下载）

(194817, 63)

In [42]:
df_test_info2 = df_test_info1[df_test_info1["knowledge_ids"].str.contains("32753")]
df_test_info2.shape
# 32753	计算树3.0人教版（勿标  可下载）

(388, 63)

In [43]:
df_test_info2 = df_test_info1[df_test_info1["knowledge_ids"].str.contains("32754")]
df_test_info2.shape
# 32754	课内数学体系

(2917, 63)

In [44]:
df_test_info2 = df_test_info1[df_test_info1["knowledge_ids"].str.contains("34731")]
df_test_info2.shape

(1, 63)

In [45]:
df_test_info2 = df_test_info1[df_test_info1["knowledge_ids"].str.contains("35878")]
df_test_info2.shape

(0, 63)

In [47]:
df_test_info2 = df_test_info1[df_test_info1["knowledge_ids"].str.contains("13293")]
df_test_info2.shape


(4, 63)

In [48]:
df_test_info2 = df_test_info1[df_test_info1["knowledge_ids"].str.contains("2,13293")]
df_test_info2.shape

(0, 63)

In [50]:
df_test_info1.shape

(202509, 63)

In [51]:
df_test_info1[["knowledge_ids"]]

(202509, 1)

# 知识点拆分

In [52]:
df_knowledge_trees[["id","parent_id","rank","rank_string"]]

id  parent_id  rank                  rank_string
0         2          0     1                          .2.
1      1173          2     2                     .2.1173.
2      1174       1173     3                .2.1173.1174.
3      1176       1173     3                .2.1173.1176.
4      1183       1174     4           .2.1173.1174.1183.
...     ...        ...   ...                          ...
4464  36723      33314     5  .2.32754.32756.33314.36723.
4465  36724      33314     5  .2.32754.32756.33314.36724.
4466  36725      33314     5  .2.32754.32756.33314.36725.
4467  36726      33314     5  .2.32754.32756.33314.36726.
4468  36727      33314     5  .2.32754.32756.33314.36727.

[4469 rows x 4 columns]

In [53]:
df_test_info[["test_id","knowledge_ids","knowledge_num"]]

test_id                                      knowledge_ids  \
0       2000200         ,3206,1173,3200,3865,3864,3207,2,3264,3199   
1       2000201                        ,1173,3344,2,3407,3339,3418   
2       2000202  ,3215,3203,7090,7091,2,7104,3306,1173,3339,331...   
3       2000203                       ,3519,1173,13448,3505,2,3732   
4       2000204                        ,3387,1173,3394,3343,2,3339   
...         ...                                                ...   
231886  2499189                             ,1198,1174,1173,1207,2   
231887  2499190                   ,1198,1210,1174,1173,1239,1229,2   
231888  2499397  ,33558,13797,13796,33557,33684,33687,2,32754,1...   
231889  2499514  ,13797,13796,33557,32758,2,32754,13808,16231,1...   
231890  2499784  ,13797,13796,33557,32758,2,32754,13808,16231,1...   

        knowledge_num  
0                 2.0  
1                 1.0  
2                 2.0  
3                 2.0  
4                 1.0  
...               ...  
231886            1.0  
231887            1.0  
231888            4.0  
231889            2.0  
231890            2.0  

[231891 rows x 3 columns]

In [54]:
df_test_info[["test_id","knowledge_ids","knowledge_num"]].dtypes

test_id            int64
knowledge_ids     object
knowledge_num    float64
dtype: object

In [ ]:
def knowledge_num_1(str1):
#     str1 = ",1176,1173,2788,2,10492"
    str1_list = [int(i) for i in str1.split(",") if i is not ""]
    str1_list.sort()
    str1_result = ".".join( [str(i) for i in str1_list])
    str1_result = "." + str1_result + "."
    return str1_result

In [ ]:
def knowledge_num_2(str2):
    pass

# get_knowledge_ids
## 拆分知识点的思路：
+ 1、将knowledge_ids转成列表(原始列表)进行排序，选出做最大值。
+ 2、选出最大值，最大值在知识树中对应的rank_string，将rank_string转成列表。
+ 3、对原始列表和rank_strin列表取差集，差集再取最大值。
+ 4、差集最大值选择知识树选择对应的rank_string_new，将rank_string_new转成列表。
+ 5、将获得的rank_string、rank_string_new等进行合集，与原始序列进行差集，取最大值
+ 6、停止条件是差集为空。


In [101]:
def get_knowledge_ids(knowledge_ids):
    """
    拆分知识点
    """
    knowledge_ids = str(knowledge_ids)
    if len(knowledge_ids)>0 :
        knowledge_ids_list = [int(i) for i in knowledge_ids.split(",") if i is not ""]
        knowledge_ids_list.sort()

        rank_string = df_knowledge_trees[df_knowledge_trees["id"] == max(knowledge_ids_list)]["rank_string"].values[0]
        rank_string_list = [rank_string]
        rank_string = [int(i) for i in rank_string.split(".") if i is not ""]

        diff = list(set(knowledge_ids_list).difference(set(rank_string)))

        while len(diff)>0:
            rank_string_new = df_knowledge_trees[df_knowledge_trees["id"] == max(diff)]["rank_string"].values[0]
            rank_string_list.append(rank_string_new)

            rank_string_new = [int(i) for i in rank_string_new.split(".") if i is not ""]
            rank_string = list(set(rank_string).union(set(rank_string_new)))
            diff = list(set(knowledge_ids_list).difference(set(rank_string)))

        return rank_string_list[::-1]
    else:
        return

In [107]:
# knowledge_ids = ",3206,1173,3200,3865,3864,3207,2,3264,3199"
knowledge_ids = ",16217,27686,13797,13796,32761,27893,16218,33824,2,33775,13804,32754,27513,27447"
get_knowledge_ids(knowledge_ids)

['.2.32754.32761.33775.33824.',
 '.2.27447.27513.27686.27893.',
 '.2.13796.13797.13804.16218.',
 '.2.13796.13797.13804.16217.']

In [108]:
knowledge_ids = ""
s = get_knowledge_ids(knowledge_ids)
print(type(s))

<class 'NoneType'>


In [156]:
df_test_info_3 = df_test_info[["test_id","knowledge_ids","knowledge_num"]][:20]

In [157]:
df_test_info_3

test_id                                      knowledge_ids  knowledge_num
0   2000200         ,3206,1173,3200,3865,3864,3207,2,3264,3199            2.0
1   2000201                        ,1173,3344,2,3407,3339,3418            1.0
2   2000202  ,3215,3203,7090,7091,2,7104,3306,1173,3339,331...            2.0
3   2000203                       ,3519,1173,13448,3505,2,3732            2.0
4   2000204                        ,3387,1173,3394,3343,2,3339            1.0
5   2000205         ,1176,3338,1173,2992,2993,2994,3004,2,2785            3.0
6   2000206         ,1173,7090,7091,3344,2,7104,3407,3339,3418            2.0
7   2000207    ,3508,1173,3541,3595,3598,3505,3506,3524,2,3537            2.0
8   2000208  ,3207,3249,1173,3259,3200,3206,13395,3263,2,13...            4.0
9   2000209                             ,1176,2901,1173,2,2794            1.0
10  2000210                            ,1176,1173,2788,2,10492            1.0
11  2000211             ,3205,1173,3200,3238,11029,3245,2,3199            2.0
12  2000212           ,1176,1173,10971,2789,2,10974,10970,2975            2.0
13  2000213                        ,1198,1174,1173,1349,2,1227            1.0
14  2000214    ,1198,1210,1174,1173,7070,1183,1366,2,7037,7038            2.0
15  2000215  ,1176,4389,11181,1173,2884,11183,11182,2,2886,...            3.0
16  2000314                                                NaN            0.0
17  2000315                                                NaN            0.0
18  2000316                                                NaN            0.0
19  2000317                                                NaN            0.0

In [146]:
df_test_info_3 = df_test_info_3[-df_test_info_3["knowledge_ids"].isna()]
df_test_info_3

In [195]:

def get_knowledge_ids(knowledge_ids):
    """
    拆分知识点
    """
    knowledge_ids = str(knowledge_ids)
    if len(knowledge_ids)>0 and knowledge_ids is not np.NAN:
        
        knowledge_ids_list = [int(i) for i in knowledge_ids.split(",") if i is not ""]
#         knowledge_ids_list = [int(i) for i in knowledge_ids.split(",") if i is not np.NAN]
     
        knowledge_ids_list.sort()

        rank_string = df_knowledge_trees[df_knowledge_trees["id"] == max(knowledge_ids_list)]["rank_string"].values[0]
        rank_string_list = [rank_string]
        rank_string = [int(i) for i in rank_string.split(".") if i is not ""]

        diff = list(set(knowledge_ids_list).difference(set(rank_string)))

        while len(diff)>0:
            rank_string_new = df_knowledge_trees[df_knowledge_trees["id"] == max(diff)]["rank_string"].values[0]
            rank_string_list.append(rank_string_new)

            rank_string_new = [int(i) for i in rank_string_new.split(".") if i is not ""]
            rank_string = list(set(rank_string).union(set(rank_string_new)))
            diff = list(set(knowledge_ids_list).difference(set(rank_string)))

        return rank_string_list[::-1]
    else:
        return

In [183]:
df_test_info_3["chai"] = df_test_info_3["knowledge_ids"].apply(get_knowledge_ids)
df_test_info_3

ValueError: max() arg is an empty sequence

In [189]:
knowledge_ids = ",16213,32801,13797,13796,32843,32845,2,32754,32755,13818,32824,32830,13803"
get_knowledge_ids(knowledge_ids)

['.2.32754.32755.32801.32845.',
 '.2.32754.32755.32801.32843.',
 '.2.32754.32755.32801.32830.',
 '.2.32754.32755.32801.32824.',
 '.2.13796.13797.13803.16213.',
 '.2.13796.13797.13803.13818.']

['.2.32754.32755.32801.32845.',#根据读法写出数字（带零和不带零）  
 '.2.32754.32755.32801.32843.',# 认识“亿"   
 '.2.32754.32755.32801.32830.',#根据准确数求近似数   
 '.2.32754.32755.32801.32824.',#改写成以“万”为单位的数   
 '.2.13796.13797.13803.16213.',#亿以上数的写法和改写   
 '.2.13796.13797.13803.13818.']#亿以上数的认识和读法     
 .2.13796.13797.13803. 大数的认识  
 .2.32754.32755.32801. 大数的认识

In [196]:
knowledge_ids = ",33719,16135,13796,33542,13908,32757,33717,13906,33720,33523,33560,16136,2,32754,33435,32758,16144,13801,13915"
get_knowledge_ids(knowledge_ids)

['.2.13796.13801.13906.16135.',
 '.2.13796.13801.13906.16136.',
 '.2.13796.13801.13908.13915.16144.',
 '.2.32754.32757.33435.33523.',
 '.2.32754.32757.33435.33542.',
 '.2.32754.32758.33560.33717.',
 '.2.32754.32758.33560.33719.',
 '.2.32754.32758.33560.33720.']

['.2.32754.32758.33560.33720.',#分数四则混合运算   
 '.2.32754.32758.33560.33719.',#分数除以分数（无情景）   
 '.2.32754.32758.33560.33717.',#整数除以分数（无情景）   
 '.2.32754.32757.33435.33542.',#分数乘分数计算    
 '.2.32754.32757.33435.33523.',#分数乘整数计算    
 '.2.13796.13801.13908.13915.16144.',#一个数除以分数     
 '.2.13796.13801.13906.16136.',#分数乘分数    
 '.2.13796.13801.13906.16135.']#分数乘整数    
.2.32754.32758.33560. 分数除法     
.2.32754.32757.33435. 分数乘法     
.2.13796.13801.13906. 分数乘法    
.2.13796.13801.13908. 分数除法  

In [197]:
knowledge_ids = ",1173,3344,2,3407,3339,3418"
get_knowledge_ids(knowledge_ids)
#'.2.1173.3339.3344.3407.3418.'#分解质因数

['.2.1173.3339.3344.3407.3418.']